src/arfs/arfs_transaction_uploader.ts

```diff
@@ -0,0 +1,238 @@
+
+/**
+ *  These are errors from the `/chunk` endpoint on an Arweave
+ *  node that we should never try to continue on
+ */
+const FATAL_CHUNK_UPLOAD_ERRORS = [
+	'invalid_json',
+	'chunk_too_big',
+	'data_path_too_big',
+	'offset_too_big',
+	'data_size_too_big',
+	'chunk_proof_ratio_not_attractive',
+	'invalid_proof'
+];
+
+interface ArFSTransactionUploaderConstructorParams {
+	gatewayUrl: URL;
+	transaction: Transaction;
+	maxConcurrentChunks?: number;
+	maxRetriesPerRequest?: number;
+}
+
+/**
+ *  A transaction uploader class that has been modified to handle uploading
+ *  the chunks of a transaction in parallel to the specified gateway url
+ *
+ * @example
+ *
+ *  ```ts
+ * 	await transaction.prepareChunks(transaction.data);
+ *
+ *		const transactionUploader = new ArFSTransactionUploader({
+ *			transaction,
+ *			gatewayUrl: new URL('https://arweave.net:443')
+ *		});
+ *
+ *		await transactionUploader.batchUploadChunks();
+ * ```
+ */
+export class ArFSTransactionUploader {
+	private chunkOffset = 0;
+	private txPosted = false;
+	private uploadedChunks = 0;
+	private failedRequests: Promise<AxiosResponse<unknown>>[] = [];
+
+	public get hasFailedRequests(): boolean {
+		return this.failedRequests.length > 0;
+	}
+
+	public get isComplete(): boolean {
+		return this.txPosted && this.uploadedChunks === this.totalChunks;
+	}
+
+	public get totalChunks(): number {
+		// eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+		return this.transaction.chunks!.chunks.length;
+	}
+
+	public get pctComplete(): number {
+		return Math.trunc((this.uploadedChunks / this.totalChunks) * 100);
+	}
+
+	private gatewayUrl: URL;
+	private transaction: Transaction;
+	private maxConcurrentChunks: number;
+	private maxRetriesPerRequest: number;
+
+	constructor({
+		gatewayUrl,
+		transaction,
+		maxConcurrentChunks = 32,
+		maxRetriesPerRequest = 5
+	}: ArFSTransactionUploaderConstructorParams) {
+		if (!transaction.id) {
+			throw new Error(`Transaction is not signed`);
+		}
+		if (!transaction.chunks) {
+			throw new Error(`Transaction chunks not prepared`);
+		}
+
+		this.gatewayUrl = gatewayUrl;
+		this.transaction = transaction;
+		this.maxConcurrentChunks = maxConcurrentChunks;
+		this.maxRetriesPerRequest = maxRetriesPerRequest;
+	}
+
+	/**
+	 * Uploads a transaction and all of its chunks until the upload is complete or has failed
+	 *
+	 * TODO: Convert this to a stream to report back event triggered progress
+	 *
+	 * @throws when any requests have failed beyond the maxRetries setting
+	 */
+	public async batchUploadChunks(): Promise<void> {
+		if (this.hasFailedRequests) {
+			throw new Error('Transaction upload has failed requests!');
+		}
+
+		if (!this.txPosted) {
+			await this.postTransactionHeader();
+
+			if (this.isComplete) {
+				return;
+			}
+		}
+
+		const numOfConcurrentUploadPromises = Math.min(this.totalChunks - this.chunkOffset, this.maxConcurrentChunks);
```

Comment:
Readability suggestion:
```suggestion
		const numRemainingChunks = this.totalChunks - this.chunkOffset;
		const numOfConcurrentUploadPromises = Math.min(numRemainingChunks, this.maxConcurrentChunks);
```